### PR TITLE
chore: Use nightly toolchain for check

### LINKED ISSUE #440: Discussion: Use unstable rust to run ci tool.
Currently iceberg-rust requires stable rust to build, which is reasonable as a library. But ci tools such as `cargo fmt`, `cargo clippy` are only used during ci or local development, and will not force downstream application/libary to use unstable rust. Some advanced features of these tools are only available in unstable rust, so I would propose to use unstable rust to run these tools.

### PR DESCRIPTION
**Pull request by liurenjie1024**
Close #440

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -19,25 +19,31 @@
 
 RUST_LOG = debug
 
+RUST_NIGHTLY_CHANNEL = nightly-2024-06-10
```
Comment (Xuanwo) [inline Makefile:22]: Hi, as I proposed in https://github.com/apache/iceberg-rust/issues/440, we can write `nightly-2024-06-10` in `rust-toolchain.toml`.

Comment (Xuanwo) [reply Makefile:22]: > Do you mean to use nightly to build?

Yes. We can add extra build task with `cargo +stable build` to make sure it work in stable channel.

Comment (Xuanwo) [reply Makefile:22]: Hi, my current plan is we have three rust version that need to test:

- Development: the nightly version we picked to develop. The commit could be stable or MSRV.
- MSRV: the MSRV we promised for downstream users.
- Stable: the latest stable version of rust.

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -29,6 +29,10 @@ concurrency:
   group: ${{ github.workflow }}-${{ github.ref }}-${{ github.event_name }}
   cancel-in-progress: true
 
+env:
+  rust_nightly: "nightly-2024-06-10"
```
Comment (Xuanwo) [inline .github/workflows/ci.yml:33]: We don't need to specify this as we load rust-toolchain by default.

File: .github/workflows/ci.yml [hunk 3]
```diff
@@ -29,6 +29,10 @@ concurrency:
...
+  rust_min: "1.77.1"
```
Comment (Xuanwo) [inline .github/workflows/ci.yml:34]: How about using `RUST_MSRV` for more clear?

File: .github/workflows/ci.yml [hunk 4]
```diff
@@ -38,6 +42,25 @@ jobs:
       - name: Check License Header
         uses: apache/skywalking-eyes/header@v0.6.0
 
+      - name: Install Rust ${{ env.rust_nightly }}
+        uses: dtolnay/rust-toolchain@stable
+        with:
+          toolchain: ${{ env.rust_nightly }}
+
+      - name: Install cargo-sort
+        uses: baptiste0928/cargo-install@v3
```
Comment (Xuanwo) [inline .github/workflows/ci.yml]: Does this action approved by the ASF?

File: .github/workflows/ci.yml [hunk 5]
```diff
@@ -61,14 +84,29 @@ jobs:
           - windows-latest
     steps:
       - uses: actions/checkout@v4
+
+      - name: Install Rust ${{ env.rust_min }}
```
Comment (Xuanwo) [inline .github/workflows/ci.yml]: `stable` always pre-install in CI, so we don't need to setup it up.

File: .github/workflows/publish.yml [hunk 6]
```diff
@@ -42,6 +46,14 @@ jobs:
           - "crates/catalog/rest"
     steps:
       - uses: actions/checkout@v4
+
+      - name: Install Rust ${{ env.rust_min }}
```
Comment (Xuanwo) [inline .github/workflows/publish.yml]: The same.

**Comment by Xuanwo** [comment]
Perhaps we can update the contributing guide with the versions we used.

**Review by Xuanwo — APPROVED** [review]
Thanks!

**Review by Fokko — APPROVED** [review]
Thanks for adding this @liurenjie1024 and thanks for the review @Xuanwo

File: .github/workflows/ci.yml [hunk 7]
```diff
@@ -63,8 +72,14 @@ jobs:
...
       - uses: actions/checkout@v4
+
+      - name: Setup Rust toolchain
+        uses: ./.github/actions/setup-builder
+        with:
+          rust-version: ${{ env.rust_msrv }}
+
       - name: Build
-        run: cargo build
+        run: make build
```
Comment (xxchan) [inline .github/workflows/ci.yml:82]: I don't get this. It seems we only installed stable toolchain `rust_msrv`, but not used in the build process. :eyes:

Comment (xxchan) [reply .github/workflows/ci.yml:82]: We should `cargo +${{ env.rust_msrv}} build`, or `rustup override set ${{ env.rust_msrv}}`

**Pull request merged**